### PR TITLE
[Feature] Support losses for DDPM-based models

### PR DESCRIPTION
**Pull request by LeoXing1996**
We support two critical features in `DDPMLoss`:
1. Support users to define how to collect `log_vars` by configs.
2. Support rescales loss corresponding to timesteps.

### CODE REVIEW

File: mmgen/models/losses/pixelwise_loss.py [hunk 1]
```diff
@@ -57,12 +57,13 @@ def gaussian_kld(mean_1, mean_2, logvar_1, logvar_2, base='e'):
...
             distribution.
-        logvar_1 (torch.Tensor): Log variance of the first (or the target)
+        mean_pred (torch.Tensor): Mean of the predicted (or the second)
+            distribution.
+        logvar_target (torch.Tensor): Log variance of the target (or the first)
             distribution
-        logvar_2 (torch.Tensor): Log variance of the second (or the predicted)
+        logvar_2 (torch.Tensor): Log variance of the predicted (or the second)
             distribution.
         base (str, optional): The log base of calculated KLD. Support ``'e'``
```
[the remark this answers is not in the document]
Comment (nbei) [reply mmgen/models/losses/pixelwise_loss.py:68]: We support xxx

File: mmgen/models/losses/ddpm_loss.py [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+        log_vars = dict()
+        for select_fn, loss_fn in zip(self.loss_select_fn_list,
+                                      self.loss_list):
+            mask = select_fn(timesteps)
+            if not mask.any():
+                loss_reduced = torch.zeros((1, ))
+            else:
+                loss_reduced = reduce_loss(loss[mask], reduction)
+            # remove original prefix in loss names
+            loss_term_name = loss_fn.loss_name().replace('loss_', '')
```
[the remark this answers is not in the document]
Comment (nbei) [reply mmgen/models/losses/ddpm_loss.py:444]: Can we?

File: mmgen/models/losses/ddpm_loss.py [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+    """Variational lower-bound loss for DDPM-based models.
+    In this loss, we calculate VLB of different timesteps with different
+    method. Detailly, ``DiscretizedGaussianLogLikelihoodLoss`` is used when
+    timesteps = 0 and ``GaussianKLDLoss`` at other timesteps.
+    To control the data flow for loss calculation, users should define
+    ``data_info`` and ``data_info_t_0`` for ``GaussianKLDLoss`` and
+    ``DiscretizedGaussianLogLikelihoodLoss`` respectively. If not passed
+    ``_default_data_info`` and ``_default_data_info_t_0`` would be used.
+    To be noted that, we only penalize variance in this loss term, and tensor
+    in output dict corresponding to mean would be detached.
```
[the remark this answers is not in the document]
Comment (nbei) [reply mmgen/models/losses/ddpm_loss.py]: To be noted that, we only penalize variance in this loss term, and the tensor in output dict corresponding to the predicted expectations that would be detached.

File: mmgen/models/losses/ddpm_loss.py [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+        log_cfgs = [
+            dict(type='quartile', reduction=REUCTION, prefix_name=PREFIX),
+            ...
+        ]
+
+    Args:
+        rescale_mode (str, optional): Mode of the loss rescale method.
+            Defaults to `''`.
+        rescale_cfg (dict, optional): Config of the loss rescale method.
+        log_cfgs (list[dict] | dict | optional): Configs to collect logs.
```
[the remark this answers is not in the document]
Comment (nbei) [reply mmgen/models/losses/ddpm_loss.py:90]: If so, you may specify more in the docs and give clear and straightforward guidance.

[the remark this answers is not in the document]
Comment (LeoXing1996) [reply mmgen/models/losses/ddpm_loss.py:126]: More doc about the design of log collection functions is added.
